dialects/refly/refly.thorin

```diff
@@ -0,0 +1,47 @@
+/// # The refly dialect {#refly}
+///
+/// The refly dialect allows to [reify](https://en.wikipedia.org/wiki/Reification_(computer_science)) and [reflect](https://en.wikipedia.org/wiki/Reflective_programming) Thorin's own representation.
+///
+/// [TOC]
+///
+/// ## Dependencies
+///
+/// none
+///
+/// ## Types
+///
+.ax %refly.Code: *;
+///
+/// ## Reify/Reflect
+///
+/// ### %refly.reify
+///
+/// Yields the internal representation of a Thorin expression as `%%refly.Code`.
+.ax %refly.reify: Π T: * -> T -> %refly.Code, normalize_reify;
+///
+/// ### %refly.reflect
+///
+/// Converts a `%%refly.Code` back to a Thorin expression.
+.ax %refly.reflect: Π T: * -> %refly.Code -> T, normalize_reflect;
+///
+/// ## Inspect
+///
+/// ### %refly.gid
+///
+/// Retuns the internal thorin::Def::gid of the argument.
+.ax %refly.gid: Π T: * -> T -> .Nat, normalize_gid;
```

Comment:
Isn't the gid very prone to changes? It's basically a new number after every optimization.
Can you elaborate on the use case?

Reply:
The use case is this:
```
.rule _ a + b = b + a if %refly.gid(b) < %refly.gid(a)
```
This is currently hard-coded in the normalizers of `%core` and is one of the things we may remove by the rule magic.